### PR TITLE
feat(framework) Add `StopRun` to `ExecServicer`

### PR DESCRIPTION
**Pull request by chongshenng**
Merge after
* #4639

### CODE REVIEW

File: src/py/flwr/superexec/exec_servicer_test.py [hunk 1]
```diff
@@ -104,3 +107,24 @@ def test_list_run_id(self) -> None:
...
+            input_run_id == run_id
+        ) & self.state.update_run_status(input_run_id, new_status=expected_run_status)
+
+        # Execute
+        response = self.servicer.StopRun(StopRunRequest(run_id=run_id), Mock())
+        run_state = self.state.get_run(run_id)
+
+        # Assert
+        self.assertTrue(response.success)
+        if run_state is not None:
```
Comment (jafermarq) [inline src/py/flwr/superexec/exec_servicer_test.py:129]: should we assert if `run_state` is None? and leave the check below outside an if statement? `run_state` should never be None in this test

Comment (chongshenng) [reply src/py/flwr/superexec/exec_servicer_test.py:129]: Yes, you're right that `run_state` should never be None. I'll add a check for this. This `if`-statement is resolve the `mypy` "Item "None" of "Run | None" has no attribute "status"" error.

**Review by jafermarq — DISMISSED** [review]

File: src/py/flwr/superexec/exec_servicer.py [hunk 2]
```diff
@@ -133,7 +134,24 @@ def StopRun(
...
+
+        # Exit if `run_id` not found
+        if not state.get_run(request.run_id):
+            context.abort(grpc.StatusCode.NOT_FOUND, "Run ID not found")
+
+        run_status = state.get_run_status({request.run_id})[request.run_id]
+
+        if run_status.status == Status.FINISHED:
+            log(ERROR, "Run ID `%s` is already finished", request.run_id)
+            return StopRunResponse(success=False)
```
Comment (danieljanes) [inline src/py/flwr/superexec/exec_servicer.py:147]: `context.abort` with an appropriate status code might be better here

File: src/py/flwr/superexec/exec_servicer.py [hunk 3]
```diff
@@ -133,7 +134,24 @@ def StopRun(
...
+
+        return StopRunResponse(
+            success=state.update_run_status(
+                run_id=request.run_id,
+                new_status=RunStatus(Status.FINISHED, SubStatus.STOPPED, ""),
+            )
```
Comment (danieljanes) [inline src/py/flwr/superexec/exec_servicer.py:153]: Let's pull out the `state.update_run_status` call to make it a bit easier to read sequentially

File: src/py/flwr/superexec/exec_servicer.py (hunk 2 above)
Comment (chongshenng) [reply src/py/flwr/superexec/exec_servicer.py]: Fixed.

File: src/py/flwr/superexec/exec_servicer.py (hunk 3 above)
Comment (chongshenng) [reply src/py/flwr/superexec/exec_servicer.py]: Fixed.

File: src/py/flwr/superexec/exec_servicer.py [hunk 4]
```diff
@@ -133,7 +134,28 @@ def StopRun(
...
+                grpc.StatusCode.FAILED_PRECONDITION,
+                f"Run ID {request.run_id} is already finished",
+            )
+
+        update_success = state.update_run_status(
+            run_id=request.run_id,
+            new_status=RunStatus(Status.FINISHED, SubStatus.STOPPED, ""),
+        )
+
+        return StopRunResponse(success=update_success)
```
Comment (danieljanes) [inline src/py/flwr/superexec/exec_servicer.py:158]:
```suggestion
        run_status = state.get_run_status({request.run_id})[request.run_id]
        if run_status.status == Status.FINISHED:
            context.abort(
                grpc.StatusCode.FAILED_PRECONDITION,
                f"Run ID {request.run_id} is already finished",
            )

        update_success = state.update_run_status(
            run_id=request.run_id,
            new_status=RunStatus(Status.FINISHED, SubStatus.STOPPED, ""),
        )
        return StopRunResponse(success=update_success)
```

File: src/py/flwr/superexec/exec_servicer_test.py [hunk 5]
```diff
@@ -104,3 +107,25 @@ def test_list_run_id(self) -> None:
         # Assert
         self.assertLess(abs(retrieved_timestamp - now().timestamp()), 1e-3)
         self.assertEqual(set(response.run_dict.keys()), {run_id})
+
+    def test_stop_run(self) -> None:
+        """Test StopRun method of ExecServicer."""
+        # Prepare
+        run_id = self.state.create_run(
+            "mock fabid", "mock fabver", "fake hash", {}, ConfigsRecord()
```
Comment (danieljanes) [inline src/py/flwr/superexec/exec_servicer_test.py:115]:
```suggestion
            "mock_fabid", "mock_fabver", "fake_hash", {}, ConfigsRecord()
```

**Review by danieljanes — APPROVED** [review]

**Pull request merged**